File: SmartTrade/web/assets/js/my-account.js
//window.onload = function () {
//    alert("ok");
//};

window.addEventListener("load", async function () {

    loadUserData();
    loadCityData()

});

async function loadUserData() {

    const response = await fetch("MyAccount");

    if (response.ok) {

        const responseJson = await response.json();

        document.getElementById("firstName").value = responseJson.fname;
        document.getElementById("lastName").value = responseJson.lname;
        document.getElementById("since").innerHTML = `Smart Trade Member Since, ${responseJson.since}`;
        document.getElementById("currentPassword").value = responseJson.password;
        document.getElementById("username").innerHTML = `Hello, ${responseJson.fname} ${responseJson.lname}`;

    }

}

async function loadCityData() {
    try {
        const response = await fetch("CityData");

        if (response.ok) {
            const responseJson = await response.json();
            const citySelect = document.getElementById("citySelect");

            responseJson.forEach(city => {
                const option = document.createElement("option");
                option.innerHTML = city.name;
                option.value = city.id;
                citySelect.appendChild(option);
            });
        } else {
            console.error("Failed to load city data: " + response.status);
        }
    } catch (error) {
        console.error("Error fetching city data:", error);
    }
}

async function saveChanges() {

    const fname = document.getElementById("firstName").value;
    const lname = document.getElementById("lastName").value;
    const currentPassword = document.getElementById("currentPassword").value;
    const newPassword = document.getElementById("newPassword").value;
    const confirmNewPassword = document.getElementById("confirmNewPassword").value;
    const citySelect = document.getElementById("citySelect").value;
    const lineOne = document.getElementById("lineOne").value;
    const lineTwo = document.getElementById("lineTwo").value;
    const postalCode = document.getElementById("postalCode").value;

    const userData = {
        fname: fname,
        lname: lname,
        currentPassword: currentPassword,
        newPassword: newPassword,
        confirmNewPassword: confirmNewPassword,
        citySelect: citySelect,
        lineOne: lineOne,
        lineTwo: lineTwo,
        postalCode: postalCode
    };

    const userDataJson = JSON.stringify(userData);

    const response = await fetch(
            "MyAccount",
            {
                method: "PUT",
                body: userDataJson,
                headers: {
                    "Content-Type": "application/json"
                }

            }
    );
    
    if(response.ok){
        
    }else{
        
    }

}

File: SmartTrade/src/java/controller/MyAccount.java
package controller;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import hibernate.Address;
import hibernate.City;
import hibernate.HibernateUtil;
import hibernate.User;
import java.io.IOException;
import java.io.PrintWriter;
import java.text.SimpleDateFormat;
import javax.servlet.ServletException;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;
import org.hibernate.Criteria;
import org.hibernate.Session;
import org.hibernate.SessionFactory;
import org.hibernate.criterion.Restrictions;

@WebServlet(name = "MyAccount", urlPatterns = {"/MyAccount"})
public class MyAccount extends HttpServlet {

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {

        HttpSession httpSession = request.getSession(false);

        if (httpSession != null && httpSession.getAttribute("user") != null) {
            User user = (User) httpSession.getAttribute("user");

            JsonObject responseObject = new JsonObject();

            String first_name = user.getFirst_name();
            String last_name = user.getLast_name();
            String password = user.getPassword();

            String since = new SimpleDateFormat("MMM yyyy").format(user.getCreated_at());

            responseObject.addProperty("fname", first_name);
            responseObject.addProperty("lname", last_name);
            responseObject.addProperty("password", password);
            responseObject.addProperty("since", since);

            Gson gson = new Gson();
            String responseText = gson.toJson(responseObject);
            response.setContentType("application/json");
            response.getWriter().write(responseText);
        }

    }

    @Override
    protected void doPut(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {

        Gson gson = new Gson();
        JsonObject userData = gson.fromJson(request.getReader(), JsonObject.class);

        String fname = userData.get("fname").getAsString();
        String lname = userData.get("fname").getAsString();
        String currentPassword = userData.get("currentPassword").getAsString();
        String newPassword = userData.get("newPassword").getAsString();
        String confirmNewPassword = userData.get("confirmNewPassword").getAsString();
        int cityId = userData.get("citySelect").getAsInt();
        String lineOne = userData.get("lineOne").getAsString();
        String lineTwo = userData.get("lineTwo").getAsString();
        String postalCode = userData.get("postalCode").getAsString();

        JsonObject responseObject = new JsonObject();
        responseObject.addProperty("status", Boolean.FALSE);

        // validate fields
        // validate fields
        HttpSession httpSession = request.getSession();

        if (httpSession != null && httpSession.getAttribute("user") != null) {

            User user = (User) httpSession.getAttribute("user");

            SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
            Session hibernateSession = sessionFactory.openSession();

            Criteria c = hibernateSession.createCriteria(User.class);
            c.add(Restrictions.eq("email", user.getEmail()));

            if (!c.list().isEmpty()) {
                User dbUser = (User) c.list().get(0);

                dbUser.setFirst_name(fname);
                dbUser.setLast_name(lname);

                if (!confirmNewPassword.isEmpty()) {
                    dbUser.setPassword(confirmNewPassword);
                } else {
                    dbUser.setPassword(currentPassword);
                }

                City city = (City) hibernateSession.load(City.class, cityId);

                Address address = new Address();
                address.setLine_1(lineOne);
                address.setLine_2(lineTwo);
                address.setPostal_code(postalCode);
                address.setCity(city);
                address.setUser(dbUser);

                //also update the user object stored in session
                httpSession.setAttribute("user", dbUser);
                //also update the user object stored in session

                hibernateSession.merge(dbUser);
                hibernateSession.save(address);

                hibernateSession.beginTransaction().commit();
                responseObject.addProperty("status", Boolean.TRUE);
                responseObject.addProperty("message", "User profile updated successfully");
                hibernateSession.close();
                

            }
        }

        String toJson = gson.toJson(responseObject);
        response.setContentType("application/json");
        response.getWriter().write(toJson);

    }

}
